File: src/TivoToDoList.py
#!/usr/bin/env python3
import json
import logging
import os
import requests
import smtplib
import sys
import pytz
from tzlocal import get_localzone
from bs4 import BeautifulSoup
from datetime import date, datetime, timedelta
from email.mime.text import MIMEText
from typing import List, Dict

import dateutil.parser
import dateutil.tz
from libtivomind import api


def local_datetime_from_utc_string(str_utc_time: str) -> datetime:
    return (
        dateutil.parser.parse(str_utc_time)
        .replace(tzinfo=dateutil.tz.UTC)
        .astimezone(dateutil.tz.tzlocal())
    )


class EpisodeDetails(object):
    __slots__ = [
        "title",
        "subtitle",
        "description",
        "requested_start_time",
        "requested_end_time",
    ]

    @classmethod
    def from_tivo_dict(cls, ep: Dict):
        ed = EpisodeDetails()
        ed.title = ep.get("title")
        ed.subtitle = ep.get("subtitle")
        ed.description = ep.get("description")
        ed.requested_start_time = local_datetime_from_utc_string(
            ep.get("requestedStartTime")
        )
        ed.requested_end_time = local_datetime_from_utc_string(
            ep.get("requestedEndTime")
        )
        return ed

    @classmethod
    def from_tvmaze_dict(cls, ep: Dict):
        ed = EpisodeDetails()
        ed.title = ep["show"]["name"]
        ed.subtitle = ep["name"]
        ed.description = ep["show"]["summary"]
        if ed.description:
            ed.description = BeautifulSoup(ed.description, "html.parser").get_text()
        ed.requested_start_time = local_datetime_from_utc_string(ep["airstamp"])
        ed.requested_end_time = ed.requested_start_time + timedelta(
            minutes=ep["runtime"]
        )
        return ed

    def to_html(self) -> str:
        subtitle = self.subtitle or self.description or "Unknown"
        ep_length = str((self.requested_end_time - self.requested_start_time))[:-3]
        return f"{self.requested_start_time:%I:%M %p}: <b>{self.title}</b> (<i>{subtitle}</i>) [{ep_length}]"


class ToDoListRetriever(object):
    def __init__(self, config):
        self.config = config

    def _get_new_eps_by_date(
        self, new_eps: List[EpisodeDetails], start_date: date
    ) -> List[EpisodeDetails]:
        result = [ep for ep in new_eps if ep.requested_start_time.date() == start_date]
        result.sort(key=lambda ep: ep.requested_start_time)
        return result

    def get_new_episodes(self, dates: List[date]) -> Dict[date, List[EpisodeDetails]]:
        logging.info(
            f"Connecting to tivo at {self.config['tivo_ip']}:{self.config['tivo_port']} using cert at {os.path.abspath(self.config['cert_path'])}"
        )

        mind = api.Mind.new_local_session(
            cert_path=self.config["cert_path"],
            cert_password=self.config["cert_password"],
            address=self.config["tivo_ip"],
            mak=self.config["tivo_mak"],
            port=self.config["tivo_port"],
        )

        local_tz = get_localzone()
        logging.info(f"Querying tivo for to do list")
        min_date_time = (
            datetime.combine(min(dates), datetime.min.time())
            .replace(tzinfo=local_tz)
            .astimezone(pytz.utc)
        )
        max_date_time = (
            datetime.combine(max(dates), datetime.max.time())
            .replace(tzinfo=local_tz)
            .astimezone(pytz.utc)
        )
        to_do_list = mind.recording_search(
            fetch_all=True,
            filt={
                "state": ["scheduled"],
                "minStartTime": min_date_time.strftime(
                    "%Y-%m-%d %H:%M:%S"
                ),
                "maxStartTime": max_date_time.strftime(
                    "%Y-%m-%d %H:%M:%S"
                )
            },
        )

        logging.debug("Finding new episodes")
        new_eps = [
            EpisodeDetails.from_tivo_dict(ep) for ep in to_do_list if ep["isNew"]
        ]

        tvmaze_show_ids = set(config.get("tvmaze_show_ids", []))
        if tvmaze_show_ids:
            logging.info(f"Querying tvmaze for schedule for {dates}")
            for date in dates:
                tvmaze_url = f"http://api.tvmaze.com/schedule?date={date}"
                guide = requests.get(tvmaze_url).json()
                new_eps += [
                    EpisodeDetails.from_tvmaze_dict(ep)
                    for ep in guide
                    if ep.get("show", {}).get("id") in tvmaze_show_ids
                ]

        return {date: self._get_new_eps_by_date(new_eps, date) for date in dates}


if __name__ == "__main__":
    logging.basicConfig(
        format="%(asctime)s [%(levelname)s] %(message)s", level=logging.DEBUG
    )
    config_filename = "TivoToDoList.conf"
    logging.debug(f"Loading config file from {os.path.abspath(config_filename)}")
    with open(config_filename, "r") as configFile:
        config = json.loads(configFile.read())

    today = datetime.today().date()
    dates_and_labels = [(today, "Today"), (today + timedelta(days=1), "Tomorrow")]

    retriever = ToDoListRetriever(config)
    new_eps_by_date = retriever.get_new_episodes([dl[0] for dl in dates_and_labels])

    message_list = []
    for date, label in dates_and_labels:
        message_list += [f"{label}'s new episodes:"] + [
            ep.to_html() for ep in new_eps_by_date.get(date, [])
        ]
        message_list += [""]

    message = "<br/>\n".join(message_list)

    mime_message = MIMEText(message, "html")
    mime_message["Subject"] = "To do list for " + str(datetime.now().date())
    mime_message["From"] = config["smtp_name"]
    mime_message["To"] = ",".join(config["to_emails"])

    if "-d" in sys.argv:
        logging.info(f"Not sending email: {mime_message}")
    else:
        logging.info(f"Sending email: {mime_message}")
        smtp_client = smtplib.SMTP(config["smtp_server"])
        smtp_client.ehlo()
        smtp_client.starttls()
        smtp_client.login(config["smtp_user"], config["smtp_password"])
        smtp_client.sendmail(
            config["smtp_user"], config["to_emails"], mime_message.as_string()
        )
        smtp_client.close()

    logging.info("Done")

File: src/TivoToDoList.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/mail"
	"net/smtp"
	"os"
	"sort"
	"strings"
	"time"

	gophermail "gopkg.in/jpoehls/gophermail.v0"
)

type episodeDetails struct {
	Title              string
	Subtitle           string
	Description        string
	RequestedStartTime time.Time
	RequestedEndTime   time.Time
}

func getDate(dateTime time.Time) time.Time {
	return time.Date(
		dateTime.Year(),
		dateTime.Month(),
		dateTime.Day(),
		0, 0, 0, 0,
		dateTime.Location())
}

func parseUtcAsLocalTime(utcTime string) time.Time {
	if utcTime == "" {
		return time.Time{}
	}
	result, err := time.Parse("2006-01-02 15:04:05", utcTime)
	if err != nil {
		panic(err)
	}
	return result.In(time.Now().Location())
}

func newEpisodeDetails(episodeMap map[string]interface{}) episodeDetails {
	title, _ := episodeMap["title"].(string)
	subtitle, _ := episodeMap["subtitle"].(string)
	description, _ := episodeMap["description"].(string)
	requestedStartTime, _ := episodeMap["requestedStartTime"].(string)
	requestedEndTime, _ := episodeMap["requestedEndTime"].(string)
	ep := episodeDetails{
		Title:              title,
		Subtitle:           subtitle,
		Description:        description,
		RequestedStartTime: parseUtcAsLocalTime(requestedStartTime),
		RequestedEndTime:   parseUtcAsLocalTime(requestedEndTime),
	}
	return ep
}

func coalesce(objs ...interface{}) interface{} {
	if objs == nil {
		return nil
	}
	for _, obj := range objs {
		if obj != nil {
			return obj
		}
	}
	return nil
}

func (ep episodeDetails) String() string {
	return fmt.Sprintf("%s: <b>%s</b> (<i>%s</i>) [%s]",
		ep.RequestedStartTime.Format("03:04 PM"), ep.Title, coalesce(ep.Subtitle, ep.Description, "Unknown"), strings.TrimRight(ep.RequestedEndTime.Sub(ep.RequestedStartTime).String(), "0s"))
}

func main() {
	log.SetFlags(log.LstdFlags | log.Lmicroseconds)

	nomail := flag.Bool("nomail", false, "do everything except sending the mail")
	flag.Parse()

	var err error
	var configBuf []byte
	if configBuf, err = ioutil.ReadFile("TivoToDoList.conf"); err != nil {
		panic(err)
	}
	var configParsed interface{}
	if err = json.Unmarshal(configBuf, &configParsed); err != nil {
		panic(err)
	}
	config := configParsed.(map[string]interface{})
	var tivoJSON []byte
	info, _ := os.Stat("toDoList.json")
	if info != nil && getDate(info.ModTime()) == getDate(time.Now()) {
		if tivoJSON, err = ioutil.ReadFile("toDoList.json"); err != nil {
			panic(err)
		}
	}

	if tivoJSON == nil {
		baseURL := config["kmttgBaseUrl"].(string)
		kmttgURL := baseURL + "/getToDo?tivo=Roamio"

		log.Printf("Making request to %s", kmttgURL)
		var resp *http.Response
		if resp, err = http.Get(kmttgURL); err != nil {
			panic(err)
		}
		defer resp.Body.Close()
		log.Printf("Got response with %d bytes", resp.ContentLength)
		if tivoJSON, err = ioutil.ReadAll(resp.Body); err != nil {
			panic(err)
		}
		log.Print("Read response body")
		if err = ioutil.WriteFile("toDoList.json", tivoJSON, 0644); err != nil {
			panic(err)
		}
		log.Print("Wrote response body to file")
	}

	log.Printf("Parsing response body to map")
	var tivoJSONParsed interface{}
	err = json.Unmarshal(tivoJSON, &tivoJSONParsed)
	if err != nil {
		panic(err)
	}
	tivoList := tivoJSONParsed.([]interface{})
	log.Printf("Parsed response body to map")
	var newEps []episodeDetails
	for _, ep := range tivoList {
		epMap := ep.(map[string]interface{})
		if epMap["isNew"].(bool) {
			newEps = append(newEps, newEpisodeDetails(epMap))
		}
	}
	log.Printf("Found %d new episodes", len(newEps))

	var todaysNewEps, tomorrowsNewEps []episodeDetails
	today := getDate(time.Now())
	tomorrow := today.AddDate(0, 0, 1)
	for _, ep := range newEps {
		epDate := getDate(ep.RequestedStartTime)
		if epDate == today {
			todaysNewEps = append(todaysNewEps, ep)
		} else if epDate == tomorrow {
			tomorrowsNewEps = append(tomorrowsNewEps, ep)
		}
	}

	sort.Slice(todaysNewEps, func(i, j int) bool {
		return todaysNewEps[i].RequestedStartTime.Unix() < todaysNewEps[j].RequestedStartTime.Unix()
	})
	sort.Slice(tomorrowsNewEps, func(i, j int) bool {
		return tomorrowsNewEps[i].RequestedStartTime.Unix() < tomorrowsNewEps[j].RequestedStartTime.Unix()
	})

	var messageList []string
	messageList = append(messageList, "Today's new episodes:")
	for _, ep := range todaysNewEps {
		messageList = append(messageList, ep.String())
	}

	messageList = append(messageList, "")

	messageList = append(messageList, "Tomorrow's new episodes")
	for _, ep := range tomorrowsNewEps {
		messageList = append(messageList, ep.String())
	}

	messageBody := strings.Join(messageList, "<br/>")

	m := &gophermail.Message{}
	m.SetFrom(config["smtp_name"].(string) + " <" + config["smtp_user"].(string) + ">")
	toEmails := config["to_emails"].([]interface{})
	var recipients []string
	for _, toEmail := range toEmails {
		m.AddTo(toEmail.(string))
		recipients = append(recipients, toEmail.(string))
	}

	m.Subject = "To do list for " + time.Now().Format("2006-01-02")
	m.HTMLBody = messageBody
	m.Headers = mail.Header{}
	m.Headers["Date"] = []string{time.Now().UTC().Format(time.RFC822)}

	msgBytes, err := m.Bytes()
	if err != nil {
		panic(err)
	}

	fmt.Printf("%s", msgBytes)

	if !(*nomail) {
		auth := smtp.PlainAuth("", config["smtp_user"].(string), config["smtp_password"].(string), config["smtp_host"].(string))
		err = smtp.SendMail(config["smtp_server"].(string), auth, config["smtp_name"].(string), recipients, msgBytes)
		if err != nil {
			panic(err)
		}
	}
}
